pkg/util/future/future.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+//
+
+package future
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/cockroachdb/cockroach/pkg/util/pprofutil"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/errors"
+)
+
+// Future represents a value which will become ready in the future.
+type Future[T any] struct {
+	// A bit of defensive type embedding in order to have all constructors return
+	// value type so that it is more difficult to make a mistake of returning
+	// 'nil' future.
+	*future[T]
+}
+
+type future[T any] struct {
+	joiner func() // optional function to invoke when joining.
+
+	mu struct {
+		syncutil.Mutex
+		ready     bool
+		done      atomic.Value // of chan struct{}, created lazily.
+		whenReady []func(v T, err error)
+	}
+
+	// Valid only after done channel signaled.
+	value T
+	err   error
+}
+
+// Make creates Future which will produce the value returned by the produce function.
+// Produce function runs on separate Go routine.  The passed in context should be
+// respected by produce function.
+func Make[T any](ctx context.Context, produce func(ctx context.Context) (T, error)) Future[T] {
```

Comment:
`Make` is unused and I'm not fond of the fact that it "internalizes" the goroutine (and that it introduces the `Join` method). Such a facility is IMO better kept on the stopper. Can we remove this from the PR to add only what we really need? The `Future` seems useful.

Reply:
This is gone.